### PR TITLE
ESP32: Add time module

### LINKED ISSUE #1620: ESP32: [wall]clock module
**Rationale:**
The ESP32 finally has the capability of keeping wallclock/real time, even when deep sleeping. The IDF provides wallclock time access via the standard C interface `[gs]ettimeofday()`. This should be exposed in some user-friendly manner. If we're feeling particularly brave, we could even pull in timezone data... Or, we could just stick with UTC and save our sanity.

- [ ] Implement basic get/set time function.
- [ ] Implement `gmtime()` access for nice, easy (but UTC-only) calendar info.
- [ ] Implement timezone support and `localtime()` access.

**Priority:** Medium/Low

**Dependencies:**
None?

### PR DESCRIPTION
**Pull request by x-logLT**
Fixes #1620 

- [x] This PR is for the `dev` branch rather than for `master`.
- [x] This PR is compliant with the [other contributing guidelines](https://github.com/nodemcu/nodemcu-firmware/blob/dev/CONTRIBUTING.md) as well (if not, please describe why).
- [ ] I have thoroughly tested my contribution.
- [x] The code changes are reflected in the documentation at `docs/en/*`.

New time module for manipulating system time/ calendar and controlling SNTP client.
Haven't fully tested all the functions tested and working as intended:
* initntp
* settimezone
* getlocal

### CODE REVIEW

**Comment by marcelstoer** [comment]
This one is missing documentation, see https://github.com/nodemcu/nodemcu-firmware/blob/dev/CONTRIBUTING.md#writing-documentation

File: components/modules/time.c [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+	
+	return 0;
+}
+
+static int time_initNTP(lua_State *L)
+{
+	size_t l;
+	const char *server = luaL_checklstring(L, 1, &l);
+	
+	if(l < 6)
```
Comment (devsaurus) [inline components/modules/time.c]: What is the purpose of checking the length here?
If you intend to have an optional argument I recommend to use [`luaL_optlstring()`](http://pgl.yoyo.org/luai/i/luaL_optlstring).

File: components/modules/time.c [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+{
+  struct tm *date;
+  
+  time_t now = luaL_checkint (L, 1);
+  luaL_argcheck (L, now >= 0, 1, "wrong arg range");
+
+  date = gmtime (&now);
+
+  /* construct Lua table */
+  lua_createtable (L, 0, 8);
```
Comment (devsaurus) [inline components/modules/time.c]: Should be factored into a separate function to share code with `time_getLocal()`.

File: components/modules/time.c [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+  { LSTRKEY("initntp"),         LFUNCVAL(time_initNTP)  },
+  { LSTRKEY("ntpenabled"),      LFUNCVAL(time_ntpEnabled) },
+  { LSTRKEY("ntpstop"),         LFUNCVAL(time_ntpStop) },
+  { LSTRKEY("epoch2cal"),      	LFUNCVAL(time_epoch2cal) },
+  { LSTRKEY("cal2epoch"),      	LFUNCVAL(time_cal2epoc) },
+  
+  { LNILKEY, LNILVAL }
+};
+
+NODEMCU_MODULE(TIME, "time", time_map, NULL);
```
Comment (devsaurus) [inline components/modules/time.c]: Missing newline...

**Comment by marcelstoer** [comment]
> If in doubt, use Lua style as noted in #1168 (comment).

Oh boy, scrolling through that issue made me realize I didn't keep my promise: https://github.com/nodemcu/nodemcu-firmware/issues/1168#issuecomment-201486819

**Comment by x-logLT** [comment]
Not at this moment.

**Pull request merged**